File: fraiseql-csharp/src/FraiseQL/Schema.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace FraiseQL
{
    /// <summary>
    /// Facade for schema management and minimal types export (TOML-based workflow)
    /// </summary>
    public static class Schema
    {
        private static SchemaRegistry registry = SchemaRegistry.Instance;

        /// <summary>
        /// Register a type definition
        /// </summary>
        public static void RegisterType(string name, Dictionary<string, object> fields, string description = null)
        {
            // Validate and extract scopes from fields
            var validatedFields = new Dictionary<string, object>();

            foreach (var (fieldName, fieldConfig) in fields)
            {
                var config = fieldConfig as Dictionary<string, object>;
                if (config != null)
                {
                    // Validate scope if present
                    if (config.TryGetValue("requiresScope", out var scopeObj) && scopeObj is string scope)
                    {
                        ValidateScope(scope, name, fieldName);
                    }

                    // Validate scopes array if present
                    if (config.TryGetValue("requiresScopes", out var scopesObj) && scopesObj is List<object> scopes)
                    {
                        if (scopes.Count == 0)
                        {
                            throw new InvalidOperationException(
                                $"Field {name}.{fieldName} has empty scopes array");
                        }
                        foreach (var s in scopes)
                        {
                            if (s is string scopeStr)
                            {
                                if (string.IsNullOrEmpty(scopeStr))
                                {
                                    throw new InvalidOperationException(
                                        $"Field {name}.{fieldName} has empty scope in scopes array");
                                }
                                ValidateScope(scopeStr, name, fieldName);
                            }
                        }
                    }

                    // Ensure not both scope and scopes
                    if (config.ContainsKey("requiresScope") && config.ContainsKey("requiresScopes"))
                    {
                        throw new InvalidOperationException(
                            $"Field {name}.{fieldName} cannot have both requiresScope and requiresScopes");
                    }
                }

                validatedFields[fieldName] = fieldConfig;
            }

            registry.Register(name, validatedFields, description);
        }

        /// <summary>
        /// Export minimal schema with only types (TOML workflow)
        /// </summary>
        public static string ExportTypes(bool pretty = true)
        {
            var types = new List<object>();

            foreach (var typeName in registry.GetTypeNames())
            {
                var typeInfo = registry.GetType(typeName);
                if (typeInfo == null) continue;

                var fieldsArray = new List<object>();
                foreach (var (fieldName, fieldConfig) in typeInfo.Fields)
                {
                    var config = fieldConfig as Dictionary<string, object>;
                    var fieldDict = new Dictionary<string, object>
                    {
                        { "name", fieldName },
                        { "type", config?.GetValueOrDefault("type", "String") ?? "String" },
                        { "nullable", config?.GetValueOrDefault("nullable", false) ?? false }
                    };

                    // Include scope fields if present
                    if (config?.ContainsKey("requiresScope") ?? false)
                    {
                        fieldDict["requiresScope"] = config["requiresScope"];
                    }
                    if (config?.ContainsKey("requiresScopes") ?? false)
                    {
                        fieldDict["requiresScopes"] = config["requiresScopes"];
                    }

                    fieldsArray.Add(fieldDict);
                }

                var typeObj = new Dictionary<string, object>
                {
                    { "name", typeName },
                    { "fields", fieldsArray }
                };

                if (!string.IsNullOrEmpty(typeInfo.Description))
                {
                    typeObj["description"] = typeInfo.Description;
                }

                types.Add(typeObj);
            }

            var schema = new Dictionary<string, object> { { "types", types } };

            var options = new JsonSerializerOptions
            {
                WriteIndented = pretty,
                PropertyNamingPolicy = null
            };

            return JsonSerializer.Serialize(schema, options);
        }

        /// <summary>
        /// Export minimal types to a file
        /// </summary>
        public static void ExportTypesFile(string outputPath)
        {
            try
            {
                var typesJson = ExportTypes(pretty: true);
                var directory = Path.GetDirectoryName(outputPath);

                if (!string.IsNullOrEmpty(directory) && !Directory.Exists(directory))
                {
                    Directory.CreateDirectory(directory);
                }

                File.WriteAllText(outputPath, typesJson);

                var typesCount = registry.GetTypeNames().Count();
                Console.WriteLine($"✅ Types exported to {outputPath}");
                Console.WriteLine($"   Types: {typesCount}");
                Console.WriteLine();
                Console.WriteLine("🎯 Next steps:");
                Console.WriteLine($"   1. fraiseql compile fraiseql.toml --types {outputPath}");
                Console.WriteLine("   2. This merges types with TOML configuration");
                Console.WriteLine("   3. Result: schema.compiled.json with types + all config");
            }
            catch (Exception)
            {
                throw new InvalidOperationException($"Failed to write types file: {outputPath}");
            }
        }

        /// <summary>
        /// Reset schema registry (useful for testing)
        /// </summary>
        public static void Reset()
        {
            registry.Clear();
        }

        /// <summary>
        /// Get all registered type names
        /// </summary>
        public static IEnumerable<string> GetTypeNames()
        {
            return registry.GetTypeNames();
        }

        /// <summary>
        /// Validate scope format: action:resource
        /// Valid patterns:
        /// - * (global wildcard)
        /// - action:resource (read:user.email, write:User.salary)
        /// - action:* (admin:*, read:*)
        /// </summary>
        private static void ValidateScope(string scope, string typeName, string fieldName)
        {
            if (string.IsNullOrEmpty(scope))
            {
                throw new InvalidOperationException($"Field {typeName}.{fieldName} has empty scope");
            }

            // Global wildcard is always valid
            if (scope == "*")
            {
                return;
            }

            // Must contain at least one colon
            if (!scope.Contains(":"))
            {
                throw new InvalidOperationException(
                    $"Field {typeName}.{fieldName} has invalid scope '{scope}' (missing colon)");
            }

            var parts = scope.Split(new[] { ':' }, 2);
            if (parts.Length != 2)
            {
                throw new InvalidOperationException(
                    $"Field {typeName}.{fieldName} has invalid scope '{scope}'");
            }

            var action = parts[0];
            var resource = parts[1];

            // Validate action: [a-zA-Z_][a-zA-Z0-9_]*
            if (!IsValidAction(action))
            {
                throw new InvalidOperationException(
                    $"Field {typeName}.{fieldName} has invalid action in scope '{scope}' (must be alphanumeric + underscore)");
            }

            // Validate resource: [a-zA-Z_][a-zA-Z0-9_.]*|*
            if (!IsValidResource(resource))
            {
                throw new InvalidOperationException(
                    $"Field {typeName}.{fieldName} has invalid resource in scope '{scope}' (must be alphanumeric + underscore + dot, or *)");
            }
        }

        /// <summary>
        /// Check if action matches [a-zA-Z_][a-zA-Z0-9_]*
        /// </summary>
        private static bool IsValidAction(string action)
        {
            if (string.IsNullOrEmpty(action))
            {
                return false;
            }

            // First character must be letter or underscore
            var firstChar = action[0];
            if (!char.IsLetter(firstChar) && firstChar != '_')
            {
                return false;
            }

            // Rest must be letters, digits, or underscores
            for (int i = 1; i < action.Length; i++)
            {
                var ch = action[i];
                if (!char.IsLetterOrDigit(ch) && ch != '_')
                {
                    return false;
                }
            }

            return true;
        }

        /// <summary>
        /// Check if resource matches [a-zA-Z_][a-zA-Z0-9_.]*|*
        /// </summary>
        private static bool IsValidResource(string resource)
        {
            if (resource == "*")
            {
                return true;
            }

            if (string.IsNullOrEmpty(resource))
            {
                return false;
            }

            // First character must be letter or underscore
            var firstChar = resource[0];
            if (!char.IsLetter(firstChar) && firstChar != '_')
            {
                return false;
            }

            // Rest must be letters, digits, underscores, or dots
            for (int i = 1; i < resource.Length; i++)
            {
                var ch = resource[i];
                if (!char.IsLetterOrDigit(ch) && ch != '_' && ch != '.')
                {
                    return false;
                }
            }

            return true;
        }
    }

    /// <summary>
    /// Central registry for GraphQL type definitions
    /// </summary>
    public class SchemaRegistry
    {
        private static SchemaRegistry instance;
        private static object lockObj = new object();
        private Dictionary<string, (Dictionary<string, object> Fields, string Description)> types = new();

        public static SchemaRegistry Instance
        {
            get
            {
                if (instance == null)
                {
                    lock (lockObj)
                    {
                        if (instance == null)
                        {
                            instance = new SchemaRegistry();
                        }
                    }
                }
                return instance;
            }
        }

        public void Register(string name, Dictionary<string, object> fields, string description = null)
        {
            lock (lockObj)
            {
                types[name] = (fields, description);
            }
        }

        public (Dictionary<string, object> Fields, string Description)? GetType(string name)
        {
            lock (lockObj)
            {
                return types.ContainsKey(name) ? types[name] : null;
            }
        }

        public IEnumerable<string> GetTypeNames()
        {
            lock (lockObj)
            {
                return types.Keys.ToList();
            }
        }

        public void Clear()
        {
            lock (lockObj)
            {
                types.Clear();
            }
        }
    }
}

File: fraiseql-ruby/lib/fraiseql/schema.rb
require 'json'
require_relative 'registry'
require_relative 'types'

module FraiseQL
  # Schema management and type export for TOML-based workflow
  class Schema
    # Export types to minimal JSON (TOML workflow)
    # The pretty parameter controls formatting (true = pretty-printed, false = compact)
    def self.export_types(pretty = true)
      registry = SchemaRegistry.instance
      types = registry.get_types

      # Convert FieldInfo objects to hashes for JSON export
      exported_types = types.map do |type_def|
        {
          name: type_def[:name],
          fields: type_def[:fields].map { |field| field.to_h },
          description: type_def[:description]
        }.tap { |h| h.delete(:description) if h[:description].nil? }
      end

      # Build minimal schema with only types
      minimal_schema = {
        types: exported_types
      }

      if pretty
        JSON.pretty_generate(minimal_schema)
      else
        JSON.generate(minimal_schema)
      end
    end

    # Export types to a file
    def self.export_types_file(output_path)
      types_json = export_types(true)

      File.write(output_path, types_json)

      # Print summary
      registry = SchemaRegistry.instance
      types_count = registry.get_types.length

      puts "✅ Types exported to #{output_path}"
      puts "   Types: #{types_count}"
      puts ""
      puts "🎯 Next steps:"
      puts "   1. fraiseql compile fraiseql.toml --types #{output_path}"
      puts "   2. This merges types with TOML configuration"
      puts "   3. Result: schema.compiled.json with types + all config"
    end

    # Register a type
    def self.register_type(name, fields_array, description = nil)
      # Validate and extract scopes from fields
      validated_fields = fields_array.map do |field_config|
        validated_field = field_config.dup

        # Validate scope if present
        if field_config[:requires_scope]
          validate_scope(field_config[:requires_scope], name, field_config[:name])
        end

        # Validate scopes array if present
        if field_config[:requires_scopes]
          if field_config[:requires_scopes].empty?
            raise RuntimeError, "Field #{name}.#{field_config[:name]} has empty scopes array"
          end
          field_config[:requires_scopes].each do |scope|
            if scope.empty?
              raise RuntimeError, "Field #{name}.#{field_config[:name]} has empty scope in scopes array"
            end
            validate_scope(scope, name, field_config[:name])
          end
        end

        # Ensure not both scope and scopes
        if field_config[:requires_scope] && field_config[:requires_scopes]
          raise RuntimeError, "Field #{name}.#{field_config[:name]} cannot have both requires_scope and requires_scopes"
        end

        validated_field
      end

      registry = SchemaRegistry.instance
      registry.register_type(name, validated_fields, description)
    end

    # Reset registry (useful for testing)
    def self.reset
      registry = SchemaRegistry.instance
      registry.reset
    end

    # Get all registered types
    def self.get_types
      registry = SchemaRegistry.instance
      registry.get_types
    end

    private

    # Validate scope format: action:resource
    # Valid patterns:
    # - * (global wildcard)
    # - action:resource (read:user.email, write:User.salary)
    # - action:* (admin:*, read:*)
    def self.validate_scope(scope, type_name, field_name)
      if scope.empty?
        raise RuntimeError, "Field #{type_name}.#{field_name} has empty scope"
      end

      # Global wildcard is always valid
      return if scope == '*'

      # Must contain at least one colon
      unless scope.include?(':')
        raise RuntimeError, "Field #{type_name}.#{field_name} has invalid scope '#{scope}' (missing colon)"
      end

      parts = scope.split(':', 2)
      if parts.size != 2
        raise RuntimeError, "Field #{type_name}.#{field_name} has invalid scope '#{scope}'"
      end

      action = parts[0]
      resource = parts[1]

      # Validate action: [a-zA-Z_][a-zA-Z0-9_]*
      unless valid_action?(action)
        raise RuntimeError, "Field #{type_name}.#{field_name} has invalid action in scope '#{scope}' (must be alphanumeric + underscore)"
      end

      # Validate resource: [a-zA-Z_][a-zA-Z0-9_.]*|*
      unless valid_resource?(resource)
        raise RuntimeError, "Field #{type_name}.#{field_name} has invalid resource in scope '#{scope}' (must be alphanumeric + underscore + dot, or *)"
      end
    end

    # Check if action matches [a-zA-Z_][a-zA-Z0-9_]*
    def self.valid_action?(action)
      return false if action.empty?

      # First character must be letter or underscore
      first_char = action[0]
      return false unless first_char.match?(/[a-zA-Z_]/)

      # Rest must be letters, digits, or underscores
      action[1..-1].each_char do |ch|
        return false unless ch.match?(/[a-zA-Z0-9_]/)
      end

      true
    end

    # Check if resource matches [a-zA-Z_][a-zA-Z0-9_.]*|*
    def self.valid_resource?(resource)
      return true if resource == '*'
      return false if resource.empty?

      # First character must be letter or underscore
      first_char = resource[0]
      return false unless first_char.match?(/[a-zA-Z_]/)

      # Rest must be letters, digits, underscores, or dots
      resource[1..-1].each_char do |ch|
        return false unless ch.match?(/[a-zA-Z0-9_.]/)
      end

      true
    end
  end
end
